doc/cpd.1
.\" $Id: cpd.1,v 1.6 2025/09/10 20:26:30 tom Exp $
.TH CPD 1 2025-09-10 MiscTools "User commands"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.de NE
.fi
.ft R
.ie n  .in -4
.el    .in -2
..
.de NS
.ie n  .sp
.el    .sp .5
.ie n  .in +4
.el    .in +2
.nf
.ft \*(CW
..
.
.hy 0
.SH NAME
cpd \-
copy-date
.SH SYNOPSIS
.B cpd
.I filename
[
.I ...
]
.
.SH DESCRIPTION
Use \fBcpd\fR to copy the timestamp from one file/directory to one or
more files/directories.
.
.SH OPTIONS
None.
.
.
.SH BUGS
.
Probably.
.
.SH EXAMPLES
.
\fBCpd\fR is used in shell scripts for a variety of reasons.
The same can be accomplished with \fBtouch\fP (though not
at the time this utility was first written).
These do the same thing now:
.
.NS 4
cpd source target
.NE
.
.PP
and
.
.NS 4
touch -r source target
.NE
.
.SH SEE ALSO
touch(1)


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/cpd/cpd.c
/*
 * $Id: cpd.c,v 1.5 2025/09/10 20:38:30 tom Exp $
 *
 * Title:	cpd.c
 * Author:	T.Dickey
 * Created:	29 Oct 1991 (refer to "copydate" script in 1982)
 * Function:	Copy file-modification time from reference file to targets.
 */
#include	<stdlib.h>
#include	<stdio.h>
#include	<string.h>
#include	<time.h>
#include	<utime.h>
#include	<sys/types.h>
#include	<sys/stat.h>

static char *leaf;
static int touch;

static void
usage(void)
{
    fprintf(stderr, "usage: %s %s dst[...]\n",
	    leaf, touch ? "date" : "src");
    exit(EXIT_FAILURE);
}

static void
wantdate(void)
{
    fprintf(stderr, "date is in form mmddhhmm[yy]\n");
    usage();
}

int
main(int argc, char *argv[])
{
    struct utimbuf timep;
    struct stat sb1;
    register int j;

    if (!(leaf = strrchr(argv[0], '/')))
	leaf = argv[0];
    touch = !strcmp(leaf, "touch");

    if (argc <= 1) {
	fprintf(stderr, "expected arguments\n");
	usage();
    }

    if (touch) {
	char *s = argv[1];
	int mon = 0, day = 0, hour = 0, min = 0, yy = 0;
	time_t when = time(NULL);
	struct tm tm;

	tm = *localtime(&when);

	switch (strlen(s)) {
	case 8:
	    yy = tm.tm_year;
	    if (sscanf(s, "%02d%02d%02d%02d",
		       &mon, &day, &hour, &min) != 4)
		wantdate();
	    break;
	case 10:
	    if (sscanf(s, "%02d%02d%02d%02d%02d",
		       &mon, &day, &hour, &min, &yy) != 5)
		wantdate();
	    break;
	default:
	    wantdate();
	    /*NOTREACHED */
	}

	tm.tm_mon = mon - 1;
	tm.tm_mday = day;
	tm.tm_hour = hour;
	tm.tm_min = min;
	tm.tm_sec = 0;
	when = mktime(&tm);
	fprintf(stderr, "touch %s", ctime(&when));
	for (j = 2; j < argc; j++) {
	    timep.actime =
		timep.modtime = when;
	    if (utime(argv[j], &timep) < 0)
		perror(argv[j]);
	}
    } else {
	for (j = 2; j < argc; j++)
	    if (stat(argv[1], &sb1) >= 0) {
		timep.actime = sb1.st_atime;
		timep.modtime = sb1.st_mtime;
		if (utime(argv[j], &timep) < 0)
		    perror(argv[j]);
	    }
    }
    return (EXIT_SUCCESS);
}
